### ISSUE TITLE
PIOT-CFG-10-001: Configure TLS for your local MQTT broker instance

### ISSUE THREAD

**Issue by labbenchstudios** [programming-the-iot/book-exercise-tasks#154]
**Description**
- This configuration activity will require enabling TLS support for your local MQTT broker instance.
  - NOTE 1: These instructions assume you're using the Mosquitto MQTT broker version 1.6.x or 2.0.x broker (running on Ubuntu 20.04 LTS), referenced below:
    - The Eclipse Mosquitto MQTT broker, written by Roger Light. References:
      - Light, (2017), Mosquitto: server and client implementation of the MQTT protocol, Journal of Open Source Software, 2(13), 265, doi:10.21105/joss.00265
      - Eclipse Foundation. Eclipse Mosquitto. (2020) [Online]. Available: https://mosquitto.org/
  - NOTE 2: Testing was done using Mosquitto 1.6.9 and 2.0.15. The latest major version of Mosquitto as of this writing is 2.0.x (see the [Mosquitto Website](https://mosquitto.org/) for details). There are a number of changes introduced with version 2.0 that you should be aware of. If you were (or are) using 1.6.x and switching to 2.0.x, be sure to review the [version migration documentation](https://mosquitto.org/documentation/migrating-to-2-0/) before beginning this exercise: https://mosquitto.org/documentation/migrating-to-2-0/

Be sure to review [PIOT-DOC-CFG](https://github.com/orgs/programming-the-iot/projects/1#card-59034829).

***Review the README***
- Please see [README.md](https://github.com/programming-the-iot/book-exercise-tasks) for further information on, and use of, this content.
- License for embedded documentation and source codes: [PIOT-DOC-LIC](https://github.com/programming-the-iot/book-exercise-tasks/issues/165)

***Estimated effort may vary greatly***
- The estimated level of effort for this exercise shown in the 'Estimate' section below is a very rough approximation. The actual level of effort may vary greatly depending on your development and test environment, experience with the requisite technologies, and many other factors.

**Actions**

NOTE: A legit certificate chain is required to enable TLS within Mosquitto and to use encrypted MQTT connections between the broker and your GDA / CDA. It follows that both your Mosquitto broker and your client applications (GDA and CDA) will need configuration updates to use TLS encryption.

- Before you start, be sure you install the Mosquitto clients (pub and sub) for easy testing (you may need to be `sudo` to execute this command:
  - `apt install mosquitto-clients`

***Generating TESTING Certificates (Use Only for Testing Purposes)***

- If you haven't created certificates before and need to generate your own, please follow the instructions listed at: https://mosquitto.org/man/mosquitto-tls-7.html.
  - Important notes to consider:
    - For testing locally, you may want to use 'localhost' as the Common Name when generating your test certificates. This will make testing consistent across your configurations, as you can just use 'localhost' for the host in your PiotConfig.props for both the GDA and CDA. Again, these certificates should be used for testing purposes only.
    - If you also want to experiment with using client certificates, keep in mind the server and client certificate organization names should be different, and you'll also need to use the appropriate hostname of your system for the fully qualified name of the host.
    - Here's a summary of the openssl commands I used to successfully implement this exercise. Note that I'm using a 2-year time period for each cert (i.e., ```-days 730```). Be sure to review the documentation listed above before proceeding.
```
openssl genrsa -out ca.key 2048
openssl req -x509 -new -nodes -key ca.key -sha256 -days 730 -out ca.crt
openssl genrsa -out server.key 2048
openssl req -new -key server.key -out server.csr
openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 730 -sha256
```
- After you generate your certificate chain using the process described in https://mosquitto.org/man/mosquitto-tls-7.html, you should have the files listed below available. These represent your CA certificate and key, server certificate and key, and optionally client certificate and key (not required for this exercise).
  - NOTE: Be sure to keep these in a safe and secure place! Your private key files should NEVER be shared and should always be stored in a secure and trusted location.
    - ca.crt
    - ca.key
    - ca.srl (may not be generated)
    - server.crt
    - server.csr
    - server.key

  - Ensure your GDA and CDA each have to the following (copy these to an appropriate path that your GDA and CDA can access, such as the 'certs' path at the top level of the GDA's and CDA's respective repositories on your local system):
    - The CA certificate file 'ca.crt'
    - The server certificate file 'server.crt'
    - If you decide to experiment with client certificates, each client app (GDA and CDA) will need its own unique private key, while the server will need access to their respective public keys (again, this is not required, and out of scope, for this exercise).
  - IMPORTANT NOTE: You should never store private keys and certificates within a non-protected repository.

***Using Mosquitto 1.6.x on Ubuntu or WSL (installed via the apt package manager - testing was done with Mosquitto version 1.6.9)***

NOTE: As indicated previously, MQTT exercises in this section have been tested using Mosquitto version 1.6.9 (running on Ubuntu 20.04 LTS and macOS Big Sur Version 11.2.3), which was installed using the instructions provided in the following cards:
  - [PIOT-CFG-06-001](https://github.com/programming-the-iot/book-exercise-tasks/issues/80)
  - [PIOT-CFG-07-001](https://github.com/programming-the-iot/book-exercise-tasks/issues/93)

- Prepare the Mosquitto 1.6 paths to support TLS using the key infrastructure you just created.
  - NOTE: Be sure to review the following before starting this process:
    - The README files located in /etc/mosquitto/ca_certificates/README and /etc/mosquitto/certs/README
    - The man page here: https://mosquitto.org/man/mosquitto-conf-5.html. Scroll to the section labeled 'Certificate based SSL/TLS Support' and 'Pre-shared-key based SSL/TLS Support'
    - The sample mosquitto.conf located here: https://github.com/eclipse/mosquitto/blob/1.6.x/mosquitto.conf
  - Ensure the Mosquitto broker has access to the files listed below. If you're using the default installation on Ubuntu, the paths are listed as follows:
    - The CA certificate file 'ca.crt': copy to {your install path}/ca_certificates
    - The server key file 'server.key': copy to {your install path}/certs
    - The server certificate file 'server.crt': copy to {your install path}/certs
  - NOTE: The configuration file example indicates {your install path} is /etc/mosquitto.

- Update the Mosquitto 1.6 configuration to support TLS using the key infrastructure you just created.
  - NOTE: If you decide to generate client certificates, be sure to copy the clients public key into {your install path}/certs and update mosquitto.conf appropriately.

IMPORTANT: Make sure your mosquitto.conf contains an import for your chosen conf.d directory. Also note that the instructions provided here are for a Linux environment and assume Mosquitto will run under the mosquitto user having appropriate privileges (e.g., sudo). You may choose to run Mosquitto as a different user, which means you'll have to place your configuration files and keys / certificates in a directory that user has permissions to access, and that the runtime user has appropriate system privileges.

- Here's a simple Mosquitto 1.6 configuration example. Note that you will likely use different paths if you want to configure Mosquitto to start automatically at system boot under a different user (i.e., your local environment may need a different important depending on where and how you configured Mosquitto):

```
# sample mosquitto.conf
persistence true
persistence_location /var/lib/mosquitto
log_dest file /var/log/mosquitto/mosquitto.log

include_dir /etc/mosquitto/conf.d
```

- As of this writing, the default configuration for Mosquitto 1.6 is located in /etc/mosquitto/mosquitto.conf and contains a config entry (include_dir, shown above) that instructs Mosquitto to look in the /etc/mosquitto/conf.d directory for custom configuration files ending in '.conf':
  - Create a file ending with '.conf' in the /etc/mosquitto/conf.d directory. Any valid filename should be OK, as long as it ends with .conf). I called mine 'tls.conf'.
  - Add the following entries (for Mosquitto 1.6)
```
# sample tls.conf

# NOTE: previous versions use 'port' instead of 'listener'
listener 8883

# use the next config entry for testing purposes only as it
# disables client authentication
allow_anonymous true

keyfile /etc/mosquitto/certs/server.key
certfile /etc/mosquitto/certs/server.crt
cafile /etc/mosquitto/ca_certificates/ca.crt
```

***Using Mosquitto 2.0.x on Ubuntu or WSL (compiled and installed via source code - testing was done with Mosquitto version 2.0.15)***

NOTE 1: As indicated previously, MQTT exercises in this section have been tested using Mosquitto version 2.0.15 (running on Ubuntu 20.04 LTS and macOS Big Sur Version 11.2.3), which was installed using the instructions provided in the following cards:
  - [PIOT-CFG-06-001](https://github.com/programming-the-iot/book-exercise-tasks/issues/80)
  - [PIOT-CFG-07-001](https://github.com/programming-the-iot/book-exercise-tasks/issues/93)

NOTE 2: In the interest of maintaining separate installs of Mosquitto (e.g., 1.6.9 and 2.0.15), the following will focus on installing Mosquitto 2.0 to your desired directory, under the mosquitto-2.0.15 path (to avoid overwriting the existing install in /etc/mosquitto - if preferred, you can simply install in /etc/mosquitto instead, although you may want to keep it in place until the new install / configuration is verified as functioning properly). For the purposes of this exercise, I'll refer to {your install path} as the path containing the mosquitto-2.0.15 directory.

E.g., {your install path} = /home/{user}/mosquitto-2.0.15

- Notes on installing Mosquitto 2.0 from source code
  - If you want to use the latest version of Mosquitto and are comfortable building from source code, you may want to download the source code and compile for your platform.
    - Download Mosquitto source: https://mosquitto.org/download/. During my testing, the file name downloaded was mosquitto-2.0.15.tar.gz
    - Review the GitHub README: https://github.com/eclipse/mosquitto
    - Build Mosquitto for WSL (note that this is an example only - the steps you need to take for your environment may vary from those stated below):
      - From the WSL command line / shell, update system and add dependencies (NOTE: On my WSL system, I only needed to install libssl-dev, libcjson1, libcjson-dev as dependencies):
        - sudo apt update
        - sudo apt upgrade
        - sudo apt install libssl-dev libcjson1 libcjson-dev
      - From the WSL command line / shell in my home directory (your selected home path may vary), create path structure, unpack source code, build, and deploy:
        - gunzip mosquitto-2.0.15.tar.gz
        - cd mosquitto-2.0.15
        - mkdir bin
        - mkdir certs
        - mkdir log
        - mkdir data
        - mkdir -p ./conf/tls
        - make
        - cp src/mosquitto ./bin
        - cp client/mosquitto_pub ./bin
        - cp client/mosquitto_sub ./bin
        - cp mosquitto.conf ./conf

- Prepare the Mosquitto 2.0 paths to support TLS using the key infrastructure you just created.
  - NOTE: Be sure to review the following before starting this process:
    - The man page here: https://mosquitto.org/man/mosquitto-conf-5.html. Scroll to the section labeled 'Certificate based SSL/TLS Support' and 'Pre-shared-key based SSL/TLS Support'
    - The sample mosquitto.conf located here: https://github.com/eclipse/mosquitto/blob/master/mosquitto.conf 
  - Ensure the Mosquitto 2.0 broker has access to the files listed below.
    - The CA certificate file 'ca.crt': copy this to {your install path}/certs
    - The server key file 'server.key': copy this to {your install path}/certs
    - The server certificate file 'server.crt': copy to {your install path}/certs
  - NOTE: The configuration file example indicates {your install path} is /etc/mosquitto.

- Update the Mosquitto 2.0 configuration to support TLS using the key infrastructure you just created.
  - NOTE: If you decide to generate client certificates, be sure to copy the clients public key into {your install path}/certs and update mosquitto.conf appropriately.

IMPORTANT: Make sure your mosquitto.conf contains the appropriate include_dir entry for your chosen conf.d directory. In my configuration, I've named this directory 'tls' to avoid confusion with the existing config directory named 'config'. You may choose to run Mosquitto as a different user, which means you'll have to place your configuration files and keys / certificates in a directory that the user has permissions to access and ensure the files Mosquitto will attempt to write into also have the appropriate permissions.

- Here's a simple Mosquitto 2.0 configuration example. Note that you will likely use different paths if you want to configure Mosquitto to start automatically at system boot under a different user (i.e., your local environment may need a different important depending on where and how you configured Mosquitto):
```
# sample mosquitto.conf
persistence true
persistence_file {your install path}/data/mosquitto.dat
log_dest file {your install path}/log/mosquitto.log
pid_file {your install path}/data/mosquitto.pid

include_dir {your install path}/conf/tls
```

- Adding the certificate links to your configuration. Note that Mosquitto will look in 'include_dir' for files ending in '.conf' to load this and other configuration information.
  - Create a file named 'tls.conf' in the {your install path}/conf/tls directory. While any valid filename should be OK provided it ends with .conf, I find it easiest to separate these into their purpose by name. Hence, I called mine 'tls.conf'.
  - Add the following entries:
```
# sample tls.conf
listener 8883

# use the next config entry for testing purposes only as it
# disables client authentication
allow_anonymous true

keyfile {your install path}/certs/server.key
certfile {your install path}/certs/server.crt
cafile {your install path}/ca.crt
```

***Update the PiotConfig.props configuration file for BOTH the GDA and CDA***

- The configuration file for both your GDA and CDA (./config/PiotConfig.props) will need to be updated to reference the newly generated server.crt file using the appropriate path. The following configuration file sample provides some guidance on how to update your `Mqtt.Gateway` section. This is further explained in [PIOT-INT-10-004](https://github.com/programming-the-iot/book-exercise-tasks/issues/114).
  - NOTE: For both the GDA and CDA, the `Mqtt.GatewayService` configuration section will generally look the same, although the 'certFile' location may need to be represented as an absolute path for your system. The example provided may not include all properties for the `Mqtt.GatewayService` - it's merely notional.

```
[Mqtt.GatewayService]
certFile       = ./cert/server.crt
host           = localhost
port           = 1883
securePort     = 8883
defaultQoS     = 0
keepAlive      = 60
enableCrypt    = True
```

NOTE 1: As mentioned previously, there are important changes in Mosquitto from ~1.6 to ~2.0. Be sure to review the migration documentation: https://mosquitto.org/documentation/migrating-to-2-0/

NOTE 2: These instructions provide a high-level summary for a very basic TLS setup using Mosquitto. They do not cover PSK - pre-shared key - configuration, or client authorization. Please review the Mosquitto documentation for details on these specific topics, as well as for more specific information on TLS configuration of Mosquitto.

**Estimate**
- Medium

**Tests**
- You'll probably want four separate console terminal windows open for this next part.
- From terminal 1: start Mosquitto - you'll probably have to run it with elevated privileges using 'sudo' and specify the configuration file that enables TLS. The example shown below assumes the command is running from WSL (or Ubuntu), with the updated Mosquitto configuration file at '{your install path}/conf/mosquitto.conf'.
```
cd {your install path}
./bin/mosquitto -c ./conf/mosquitto.conf
```
- From terminal 2: If you haven't already, install the mosquitto clients (which will allow you to issue pub / sub requests from the command line), followed by another command to tail the mosquitto log file (to see how it responds to your next set of tests - note that you may have to run the 2nd command with elevated privileges depending on your environment configuration):
```
sudo apt install mosquitto-clients
cd {your install path}
tail -f ./log/mosquitto.log
```
- From terminal 3: Start a simple subscribe request:
  - NOTE: be sure to use the same hostname / IP that you used when creating the server.crt file! If you can't remember, you can add the `--insecure` parameter to the client.
```
cd {your install path}
./bin/mosquitto_sub -h localhost -p 8883 --cafile ./certs/server.crt -t foobar
```
- From terminal 4: Execute a simple publish request:
  - NOTE: be sure to use the same hostname / IP that you used when creating the server.crt file! If you can't remember, you can add the `--insecure` parameter to the client.
```
cd {your install path}
./bin/mosquitto_pub -h localhost -p 8883 --cafile ./certs/server.crt -t foobar -m hello
```
- In terminal 3, you should see the following:
```
hello
```
- You may also want to start Wireshark, or another protocol analyzer, and track TLS messages on the loopback adapter to verify that the MQTT messages are, in fact, encrypted.
- Check the log file - it should include output similar to the following (note that my test uses Mosquitto 2.0.15, which is reflected in the log output below):
```
1661312116: mosquitto version 2.0.15 starting
1661312116: Config loaded from ../conf/mosquitto.conf.
1661312116: Opening ipv4 listen socket on port 8883.
1661312116: Opening ipv6 listen socket on port 8883.
1661312116: mosquitto version 2.0.15 running
1661312170: New connection from 127.0.0.1:55180 on port 8883.
.
.
.
1661312242: Sending CONNACK to ... (0, 0)
1661312242: Received PUBLISH from ... (d0, q0, r0, m0, 'foobar', ... (5 bytes))
1661312242: Sending PUBLISH to ... (d0, q0, r0, m0, 'foobar', ... (5 bytes))
1661312242: Received DISCONNECT from ...
```
- If you're testing with Mosquitto version 1.6.9, your log output may look similar to the following:
```
1614467321: mosquitto version 1.6.9 starting
1614467321: Config loaded from /etc/mosquitto/mosquitto.conf.
1614467321: Opening ipv4 listen socket on port 8883.
1614467321: Opening ipv6 listen socket on port 8883.
1614467321: New connection from 127.0.0.1 on port 8883.
1614467321: New client connected from 127.0.0.1 as ...
1614467770: Client ... disconnected.
1614467774: New connection from 127.0.0.1 on port 8883.
1614467774: New client connected from 127.0.0.1 as ....
1614467812: New connection from 127.0.0.1 on port 8883.
```
